Filter out short-lived LLVM diagnostics before they reach the rustc handler

lqd opened the pull request:
During profiling I saw remark passes being unconditionally enabled: for example `Machine Optimization Remark Emitter`.  

The diagnostic remarks enabled by default are [from missed optimizations and opt analyses](https://github.com/rust-lang/rust/pull/113339#discussion_r1259480303). They are created by LLVM, passed to the diagnostic handler on the C++ side, emitted to rust, where they are unpacked, C++ strings are converted to rust, etc.

Then they are discarded in the vast majority of the time (i.e. unless some kind of `-Cremark` has enabled some of these passes' output to be printed). 

These unneeded allocations are very short-lived, basically only lasting between the LLVM pass emitting them and the rust handler where they are discarded. So it doesn't hugely impact max-rss, and is only a slight reduction in instruction count (cachegrind reports a reduction between 0.3% and 0.5%) _on linux_. It's possible that targets without `jemalloc` or with a worse allocator, may optimize these less.

It is however significant in the aggregate, looking at the total number of allocated bytes:
- it's the biggest source of allocations according to dhat, on the benchmarks I've tried e.g. `syn` or `cargo`
- allocations on `syn` are reduced by 440MB, 17% (from 2440722647 bytes total, to 2030461328 bytes)
- allocations on `cargo` are reduced by 6.6GB, 19% (from 35371886402 bytes total, to 28723987743 bytes)

Some of these diagnostics objects [are allocated in LLVM](https://github.com/rust-lang/rust/pull/113339#discussion_r1252387484) *before* they're emitted to our diagnostic handler, where they'll be filtered out. So we could remove those in the future, but that will require changing a few LLVM call-sites upstream, so I left a FIXME.

lqd commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 1a8ebe489d2077d188ad02bc2af7455136b39829 with merge 86a42973b773876956b20455a549ebcb24dd9e6e...
<!-- homu: {"type":"TryBuildStarted","head_sha":"1a8ebe489d2077d188ad02bc2af7455136b39829","merge_sha":"86a42973b773876956b20455a549ebcb24dd9e6e"} -->

lqd commented on compiler/rustc_llvm/llvm-wrapper/RustWrapper.cpp:
```diff
@@ -1995,7 +1995,7 @@ extern "C" void LLVMRustContextConfigureDiagnosticHandler(
     std::move(RemarkStreamer),
     std::move(LlvmRemarkStreamer)
-  ));
+  ), true);
```
Yeah I stopped here at seeing that our diagnostics handler wasn’t called anymore, to check whether changes to the PGO diagnostics would be impactful. 

This is great point, thank you, I will look into that.

lqd replied:
Here's some of what I found. 

On debug builds for helloworld,diesel,exa,cargo,syn together (the latter 2 generate *a lot* of optimization diagnostics, we're talking 6-7GBs in total on cargo for example, that `RespectFilters = true` eliminates), there were 330k diagnostics total. 

They are coming from 2 passes, "sdagisel" the more common (260k, 80%), and "asm-printer" next (72k, 20%).

Here's the top 60% to give you an idea:

```
336578 counts
(  1)   166274 (49.4%, 49.4%): optimization diagnostic received, kind: OptimizationMissed, pass: "sdagisel", message: "FastISel missed terminator"
(  2)    23695 ( 7.0%, 56.4%): optimization diagnostic received, kind: OptimizationMissed, pass: "sdagisel", message: "FastISel didn't lower all arguments: ptr"
(  3)     7822 ( 2.3%, 58.8%): optimization diagnostic received, kind: OptimizationMissed, pass: "sdagisel", message: "FastISel missed"
(  4)     3962 ( 1.2%, 59.9%): optimization diagnostic received, kind: OptimizationAnalysis, pass: "asm-printer", message: "8 instructions in function"
...
```

The rest are all around the <=1% range, with unique fastisel messages mentioning the function itself with missed optimizations.

With filters enabled, LLVM will not do as many of such allocations, and smaller ones. IIUC they start initially small and sometimes when the pass is enabled some additional details are appended. 

In any case:

The asm printer ones are the easiest, they're just optimization analysis diagnostics with the instruction count in a function, coming from [here](https://github.com/rust-lang/llvm-project/blob/22897bce7bfedc9cd3953a33419b346936263500/llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp#L1752-L1755).

I'll point towards a couple of the missed optimizations diagnostics in selection dag isel, but they're going through [`reportFastISelFailure`](https://github.com/rust-lang/llvm-project/blob/22897bce7bfedc9cd3953a33419b346936263500/llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp#L663-L678) (so the 4 callers will be enough) just before going through `LLVMContext::diagnose` you linked above:
- the most common "FastISel missed terminator"/"FastISel missed" from [here](https://github.com/rust-lang/llvm-project/blob/22897bce7bfedc9cd3953a33419b346936263500/llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp#L1590-L1608), and this is a case where some details are added on-demand.
- "FastISel didn't lower all arguments: ..." from [here](https://github.com/rust-lang/llvm-project/blob/22897bce7bfedc9cd3953a33419b346936263500/llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp#L1407-L1413).

(also fun: these passes seem to go through `LegacyPassManager.cpp` which I wasn't expecting to still see :)

It feels like maybe we should land `RespectFilters = true` ourselves, and either open LLVM issues or PRs in the future, what do you think ?

These won't impact max-rss under a good allocator since they're so short-lived, and the improvements on instruction counts locally on cachegrind are understandably small (and on the perfbot it's mostly within noise on syn/cargo).

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/5458534581/jobs/9934360004)
Build commit: 86a42973b773876956b20455a549ebcb24dd9e6e (`86a42973b773876956b20455a549ebcb24dd9e6e`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/5458534581/jobs/9934360004"},"merge_sha":"86a42973b773876956b20455a549ebcb24dd9e6e"} -->

rust-timer commented:
Queued 86a42973b773876956b20455a549ebcb24dd9e6e with parent 9e4e58470520c72b4aecc81f54cf260fdf337176, future [comparison URL](https://perf.rust-lang.org/compare.html?start=9e4e58470520c72b4aecc81f54cf260fdf337176&end=86a42973b773876956b20455a549ebcb24dd9e6e).
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (86a42973b773876956b20455a549ebcb24dd9e6e): [comparison URL](https://perf.rust-lang.org/compare.html?start=9e4e58470520c72b4aecc81f54cf260fdf337176&end=86a42973b773876956b20455a549ebcb24dd9e6e&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=9e4e58470520c72b4aecc81f54cf260fdf337176&end=86a42973b773876956b20455a549ebcb24dd9e6e&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.6% | [-0.7%, -0.6%] | 2     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.6% | [-0.7%, -0.6%] | 2     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=9e4e58470520c72b4aecc81f54cf260fdf337176&end=86a42973b773876956b20455a549ebcb24dd9e6e&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -3.2% | [-3.2%, -3.2%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=9e4e58470520c72b4aecc81f54cf260fdf337176&end=86a42973b773876956b20455a549ebcb24dd9e6e&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=9e4e58470520c72b4aecc81f54cf260fdf337176&end=86a42973b773876956b20455a549ebcb24dd9e6e&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 653.846s -> 654.057s (0.03%)

lqd committed:
make opt diagnostic kinds printable

lqd committed:
filter LLVM diagnostics before crossing the rust bridge

this will eliminate many short-lived allocations (e.g. 20% of the memory used
building cargo) when unpacking the diagnostic and converting its various
C++ strings into rust strings, just to be filtered out most of the time.

lqd committed:
remove remark filtering on the rust side

now that remarks are filtered before cg_llvm's diagnostic handler callback
is called, we don't need to do the filtering post c++-to-rust conversion
of the diagnostic.

lqd commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 12c57024d5d15802c5b6e5a7b69a3835ed8f714b with merge 583a60dbbea5e6712b308f359519270f875975b0...
<!-- homu: {"type":"TryBuildStarted","head_sha":"12c57024d5d15802c5b6e5a7b69a3835ed8f714b","merge_sha":"583a60dbbea5e6712b308f359519270f875975b0"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/5586386914/jobs/10211905469)
Build commit: 583a60dbbea5e6712b308f359519270f875975b0 (`583a60dbbea5e6712b308f359519270f875975b0`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/5586386914/jobs/10211905469"},"merge_sha":"583a60dbbea5e6712b308f359519270f875975b0"} -->

rust-timer commented:
Queued 583a60dbbea5e6712b308f359519270f875975b0 with parent 0f16bd341eaf3ad47134c6d855a764e0099fbc93, future [comparison URL](https://perf.rust-lang.org/compare.html?start=0f16bd341eaf3ad47134c6d855a764e0099fbc93&end=583a60dbbea5e6712b308f359519270f875975b0).
There are currently 3 other artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~4.9 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (583a60dbbea5e6712b308f359519270f875975b0): [comparison URL](https://perf.rust-lang.org/compare.html?start=0f16bd341eaf3ad47134c6d855a764e0099fbc93&end=583a60dbbea5e6712b308f359519270f875975b0&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=0f16bd341eaf3ad47134c6d855a764e0099fbc93&end=583a60dbbea5e6712b308f359519270f875975b0&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.5% | [-0.5%, -0.5%] | 2     |
| Improvements ✅ <br /> (secondary) | -1.0% | [-2.2%, -0.6%] | 5     |
| All ❌✅ (primary)                 | -0.5% | [-0.5%, -0.5%] | 2     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=0f16bd341eaf3ad47134c6d855a764e0099fbc93&end=583a60dbbea5e6712b308f359519270f875975b0&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -1.6% | [-2.2%, -1.0%] | 2     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -1.6% | [-2.2%, -1.0%] | 2     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=0f16bd341eaf3ad47134c6d855a764e0099fbc93&end=583a60dbbea5e6712b308f359519270f875975b0&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.4% | [1.4%, 1.4%] | 1     |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 1.4% | [1.4%, 1.4%] | 1     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=0f16bd341eaf3ad47134c6d855a764e0099fbc93&end=583a60dbbea5e6712b308f359519270f875975b0&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 658.53s -> 657.986s (-0.08%)

lqd commented:
@tmiasko I've removed `RespectFilters = true` as we discussed, to do the filtering in our `RustDiagnosticHandler` when a pass/DI is disabled.

I've also removed the filtering done on the rust side, since that's now unused.

r? tmiasko

bors commented:
:pushpin: Commit ca5a383fb6b16679b31029ab5da8438b5ef03e1e has been approved by `tmiasko`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=tmiasko ca5a383fb6b16679b31029ab5da8438b5ef03e1e -->
<!-- homu: {"type":"Approved","sha":"ca5a383fb6b16679b31029ab5da8438b5ef03e1e","approver":"tmiasko","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit ca5a383fb6b16679b31029ab5da8438b5ef03e1e with merge f77c624c03065298ac76051a4c1c5c2cfeafc0e0...
<!-- homu: {"type":"BuildStarted","head_sha":"ca5a383fb6b16679b31029ab5da8438b5ef03e1e","merge_sha":"f77c624c03065298ac76051a4c1c5c2cfeafc0e0"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/5732779436/job/15538197550)
Approved by: tmiasko
Pushing f77c624c03065298ac76051a4c1c5c2cfeafc0e0 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"tmiasko","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/5732779436/job/15538197550"},"merge_sha":"f77c624c03065298ac76051a4c1c5c2cfeafc0e0"} -->

rust-timer commented:
Finished benchmarking commit (f77c624c03065298ac76051a4c1c5c2cfeafc0e0): [comparison URL](https://perf.rust-lang.org/compare.html?start=d12c6e947ceacf3b22c154caf9532b390d8dc88a&end=f77c624c03065298ac76051a4c1c5c2cfeafc0e0&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - ACTION NEEDED


**Next Steps**: If you can justify the regressions found in this perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please open an issue or create a new PR that fixes the regressions, add a comment linking to the newly created issue or PR, and then add the `perf-regression-triaged` label to this PR.

@rustbot label: +perf-regression
cc @rust-lang/wg-compiler-performance



### [Instruction count](https://perf.rust-lang.org/compare.html?start=d12c6e947ceacf3b22c154caf9532b390d8dc88a&end=f77c624c03065298ac76051a4c1c5c2cfeafc0e0&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.3%  | [0.3%, 0.3%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 0.4%  | [0.3%, 0.5%]   | 2     |
| Improvements ✅ <br /> (primary)   | -0.4% | [-0.6%, -0.2%] | 3     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.2% | [-0.6%, 0.3%]  | 4     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=d12c6e947ceacf3b22c154caf9532b390d8dc88a&end=f77c624c03065298ac76051a4c1c5c2cfeafc0e0&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=d12c6e947ceacf3b22c154caf9532b390d8dc88a&end=f77c624c03065298ac76051a4c1c5c2cfeafc0e0&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=d12c6e947ceacf3b22c154caf9532b390d8dc88a&end=f77c624c03065298ac76051a4c1c5c2cfeafc0e0&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: missing data

lqd commented:
The regressions are in 2 doc benchmarks that look a bit noisy right now, and this PR wouldn't impact rustdoc.

@rustbot label: +perf-regression-triaged
